Batteries/Control/AlternativeMonad.lean

```diff
@@ -0,0 +1,204 @@
+/-
+Copyright (c) 2025 Devon Tuma. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Devon Tuma
+-/
+import Batteries.Control.OptionT
+
+
+/-!
+# Laws for Monads with Failure
+
+Definitions for monads that also have an `Alternative` instance while sharing the underlying
+`Applicative` instance, and a class `LawfulAlternative` for types where the failure and monad
+structures are compatible in a natural way. More specifically they satisfy:
+
+* `failure >>= g = failure`
+* `x <|> failure = x`
+* `failure <|> y = y`
+* `x <|> y <|> z = (x <|> y) <|> z`
+* `f <$> (x <|> y) = (f <$> x <|> f <$> y)`
+
+`Option`/`OptionT` are the most basic examples, but transformers like `StateT` also preserve
+the lawfulness of this on the underlying monad.
+
+The law `do _ ← x; failure = failure` is true for monads like `Option` and `List` that don't
+have any "side effects" to execution, but not for something like `OptionT` on some monads,
+so we don't include this condition.
+
+We also define a class `LawfulAlternativeLift` similar to `LawfulMonadLift` that states that
+a lifting between monads preserves `failure` and `orElse`.
+
+## Tags
+
+monad, alternative, failure
+-/
+
+universe u v w
+
+section Alternative
+
+variable {F : Type → Type v} [Alternative F]
+
+-- [todo] add notation for `Functor.mapConst` and port `Functor.mapConstRev`
+/-- Returns `pure true` if the computation succeeds and `pure false` otherwise. -/
+def succeeds {α} (x : F α) : F Bool :=
```

Comment:
Upstreaming these seems orthogonal to the rest of the PR; I'd suggest splitting them into a separate PR, or leaving them in mathlib